### PR TITLE
Fix not supported messages for examples

### PR DESCRIPTION
**Pull request by tsherif**
The fact that `AnimationLoop.start` is asynchronous is a headache...

### CODE REVIEW

File: examples/api/texture-3d/app.js [hunk 1]
```diff
@@ -170,7 +168,7 @@ export default class AppAnimationLoop extends AnimationLoop {
   }
 
   isSupported() {
-    return this.isDemoSupported;
+    return this.demoNotSupported;
```
Comment (Pessimistress) [inline examples/api/texture-3d/app.js]: `!this.demoNotSupported`?

Is this used?

Comment (tsherif) [reply examples/api/texture-3d/app.js]: Oh no, those methods aren't used any more.

**Comment by ibgreen** [comment]
> The fact that AnimationLoop.start is asynchronous is a headache...

Some thoughts:
- Naturally, async issues either need to be handled by the app or the framework - i.e. if luma doesn't handle them, the apps must. 
- Before this method was asynchronous, the texture / model loading examples contained a bunch of loading callbacks and promise chains that cluttered up the code and made even simple luma.gl examples look rather complex.
- If you are actually considering making a change here, perhaps by adopting `async`/`await` syntax in the examples this could allow examples to remain relatively clean since that allows replacing the callbacks with  "linear" code.

**Comment by tsherif** [comment]
> Naturally, async issues either need to be handled by the app or the framework - i.e. if luma doesn't handle them, the apps must.

They still need to be handled by the app. The website is a client of luma.gl and it's made more complex because of this.  Part of the problem is that `start` doesn't even return the promise it creates, so the app has no way of chaining something to it. That's why I had to add the ugly `waitForRender` to the example scaffolding.

> Before this method was asynchronous, the texture / model loading examples contained a bunch of loading callbacks and promise chains that cluttered up the code and made even simple luma.gl examples look rather complex.

I don't think the API should be geared towards making the examples look simple at the expense of actual usage. Asynchronicity in luma.gl has lead to messiness elsewhere (e.g. Render tests having to wait for textures to load, the ScenegraphLayer's "wait for assets" logic), so there's a case to be made for at least rethinking how it's used.

**Comment by tsherif** [comment]
> this has never been about adding hacky APIs

I didn't mean to imply the API was hacky. Just that the benefits it brings for simple usage like the examples doesn't, to me, make up for the complexity it adds to real-world usage like the deck layers (and to luma.gl itself, e.g. in `Model`'s checks for async textures). It's all a question of tradeoffs, and this seems like a losing one to me.

> would appreciate a collaboration around that

For sure. I'm wondering if we could keep the core of luma.gl synchronous and just have some helper function to help with the async parts (e.g. just export `getPageLoadPromise` and let the application use it, rather than making it internal to `AnimationLoop`).

**Comment by ibgreen** [comment]
> For sure. I'm wondering if we could keep the core of luma.gl synchronous... Model's checks for async textures

I can only say that the usage of textures in deck.gl layers was quite tricky for end users before we added this feature. If you feel you have a better design that keeps the deck.gl application usage simple, go for it.

That said, I would be a little concern that this pushes even more complexity into deck.gl, and I generally like to push complexity downwards, not upwards, but I understand where you are coming from.

> just export getPageLoadPromise 

In principle, yes, however that construct a problematic one, it has some problems that has hit users badly (e.g. if the user defines the same window callback only one of them fires) so I regard that as as temporary "ugliness under the hood". I would not really advise exporting it.

**Comment by tsherif** [comment]
Just to give an explicit example, code like this is to me strong evidence of a bad abstraction: https://github.com/uber/deck.gl/blob/master/modules/mesh-layers/src/scenegraph-layer/gltf-utils.js#L3-L17

And to your point, it's the async textures that are pushing this complexity into deck rather than removing it. The way I see it, async should be kept as high as possible in an application stack because the "contagion" moves upward, i.e. any function that calls an async function must itself be async.

**Comment by ibgreen** [comment]
Agreed, that was a major hack.

It should please you learn that this code can now be cut. As of loaders.gl v2, the `GLTFLoader` now loads and decodes all assets, including textures, before resolving the returned promise. That means that there is no need to wait for anything in the application, all images are loaded and decoded and available on the `image.image` field.

I actually spent a significant amount of work on addressing this in the last development cycle (developing the required "composite loader" architecture in loaders.gl core). So it would be nice to see this taken advantage of in deck (doesn't luma.gl/experimental have a similar snippet?).

**Comment by ibgreen** [comment]
Another area where loaders.gl is trying to help with async handling is the new `loadImageCube` function in `@loaders.gl/images`

The IBL environment class should become significantly simpler and more generic if we used `loadImageCube`: https://github.com/uber/luma.gl/blob/master/modules/experimental/src/gltf/gltf-environment.js

I think this class would be quite hard to work with right now, as it hard codes expectations on the three different cube maps involved (e.g. whether they have mip levels).

**Comment by ibgreen** [comment]
> The way I see it, async should be kept as high as possible in an application stack because the "contagion" moves upward, i.e. any function that calls an async function must itself be async.

Fine, I get your point though the async API might not be the best example of async API contagion.

In fact, the whole point of the async textures design is that they provide a non-async API...  There is no async API "contagion" as there is no `async` qualifier or Promise-valued functions involved in the async Texture API. The async tracking is completely internal.

**Comment by tsherif** [comment]
> In fact, the whole point of the async textures design is that they provide a non-async API... The async tracking is completely internal.

The examples I gave demonstrate that this is simply not the case. The website, render tests, gltf loading... the attempt to hide the asynchronicity simply forced these applications to find ways to dig into the internals of luma to manage the async calls (in brittle ways). I'll put it this way: the only time you can hide async calls is if the application doesn't care about the results of those calls, and there are very few serious applications I've come across that don't care how their resources are managed.

**Comment by ibgreen** [comment]
@tsherif  Thanks for taking the time to explain your thinking in depth! It sounds like you are set on which direction to take here, which is great. 

I'll proceed with the assumption that async features will gradually be phased out from luma.gl.

**Pull request merged**